arcgis-runtime-samples-macos/Layers/List KML contents/ListKMLContentsViewController.swift

```
@@ -0,0 +1,158 @@
+        }
+    }
+    
+    /// Sets `isVisible` to `true` for these nodes and their desdendants
+    func makeNodesVisible(_ nodes: [AGSKMLNode]){
+        for node in nodes{
+            node.isVisible = true
+            makeNodesVisible(childNodes(of: node))
+        }
+    }
+    
+    /// Returns all the child nodes of this node
+    func childNodes(of node: AGSKMLNode) -> [AGSKMLNode] {
+        var children = [AGSKMLNode]()
+        if let container = node as? AGSKMLContainer {
+            children.append(contentsOf: container.childNodes)
+        }
+        if let networkLink = node as? AGSKMLNetworkLink {
+            children.append(contentsOf: networkLink.childNodes)
+        }
+        return children
+    }
+
+    /// Sets the viewpoint of the scene to the extent of the node, if it has one
+    func setSceneViewpoint(on node: AGSKMLNode){
+        if let extent = node.extent,
+            // some nodes do not include a geometry, so check that the extent isn't empty
+            !extent.isEmpty {
+           
+            let viewpoint = AGSViewpoint(targetExtent: extent)
+            sceneView.setViewpoint(viewpoint)
+        }
+    }
+    
+}
+
+// Boilerplate datasource code to display the KML node hierarchy
+extension ListKMLContentsViewController: NSOutlineViewDataSource {
+    
+    func outlineView(_ outlineView: NSOutlineView, numberOfChildrenOfItem item: Any?) -> Int {
+        if let node = item as? AGSKMLNode {
+            return childNodes(of: node).count
+        }
+        else if let kmlDataset = kmlDataset {
+            return kmlDataset.rootNodes.count
+        }
+        return 0
+    }
+    
+    func outlineView(_ outlineView: NSOutlineView, child index: Int, ofItem item: Any?) -> Any {
+        if let node = item as? AGSKMLNode {
+            return childNodes(of: node)[index]
+        }
+        else if let kmlDataset = kmlDataset {
+            return kmlDataset.rootNodes[index]
+        }
+        return NSNull()
```

This is essentially a never-should-happen case, right? If so, I recommend just using `else { fatalError() }`. That way if the won't ever happen does happen as a result of a development change, it will be caught immediately.

---

I try never to use `fatalError` unless there's a reason the whole app cannot continue. If that case ever did happen, I'd rather have a blank outline view in one sample than crash entirely.

---

I hear what you're saying. And it is rather scary to see `fatalError()` in production code. But just because it's there doesn't mean it will be be executed. I'd rather the app crash on us (as developers) to tell us that we've done something wrong rather than have a silent failure go overlooked. Do you know how `NSOutlineView` handles `NSNull`? It might just fail anyway.

---

My experience with `NSOutlineView` is fairly limited, but based on what I know, this is somewhat analogous to `cellForRowAtIndexPath`. The data source has already said that there's a data model object at the given index. That object is either a top level object or a child object. Returning `NSNull` would be the equivalent of returning `nil` (if the API would let you) when a cell is requested. That's crazy! The data source is essentially saying that there is and there is not something there.

From the documentation for `outlineView(_:child:ofItem:):

> The child item at index of item. If item is nil, returns the appropriate child item of the root object.

There is no `nil` option. That's a big advantage I see to using a separate object initialized with a non-`nil` data set. The data set would never be `nil`, preventing that scenario altogether:
```swift
if let node = item as? AGSKMLNode {
    return childNodes(of: node)[index]
} else {
    return dataSet.rootNodes[index]
}
```

---

I changed it so the data source is set after the data is loaded, so that `kmlDataset` can never be `nil` and the data source functions never need to account for a `nil` case.